### PR TITLE
Update Azure Configuration Service to use Builder.

### PR DESCRIPTION
**Pull request by conniey**
* Updates ConfigurationClient to use a builder for creation
* Deletes ApplicationClientCredentials and moves logic to ConfigurationClientCredentials, which now inherits from AsyncServiceClientCredentials
* Updates spotbugs pom.xml to use new v3 package names.

Relies on:
https://github.com/Azure/autorest-clientruntime-for-java/pull/592
https://github.com/Azure/autorest-clientruntime-for-java/pull/593

### CODE REVIEW

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java [hunk 1]
```diff
@@ -60,36 +60,32 @@
...
 
-            List<HttpPipelinePolicy> policies = ConfigurationClient.getDefaultPolicies(connectionString);
-            policies.add(interceptorManager.getRecordPolicy());
-            policies.add(loggingPolicy);
-
-            pipeline = new HttpPipeline(HttpClient.createDefault().wiretap(true), policies);
+            client = ConfigurationClient.builder(new ConfigurationClientCredentials(connectionString))
+                    .withHttpClient(HttpClient.createDefault().wiretap(true))
+                    .withHttpLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS)
+                    .withPolicy(interceptorManager.getRecordPolicy())
```
Comment (JonathanGiles) [inline applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: We should discuss the 'with' naming pattern. I've observed within Microsoft that two patterns appear to exist for fluent API, `withPolicy(..)` and `policy(..)`.

Comment (conniey) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: Is there one that is more predominantly used? I went with the naming that I saw in a random class here: https://github.com/Azure/azure-libraries-for-java/

Comment (JonathanGiles) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: So funnily enough they are pretty much the 'land of the with' :-) I would say that from what I've seen closer to our home in azure-sdk-for-java we typically don't use 'with'. I would like to get any input from @anuchandy and @jianghaolu though, because theoretically they straddle both sides now and I'm not clear why this discrepancy arose.

If it were me, well, I come from the land where everything should be a standard, non-fluent setter method :-) I would probably say without 'with' is right for us, as that is what we see in azure-common today.

Comment (JonathanGiles) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: But please disagree with me. I hold almost no opinion here and want what is right.

Comment (JonathanGiles) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: I've done more research and I now feel more opinionated :-) I don't like the 'with' naming pattern.

Comment (conniey) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: I don't have an opinion. What was the rationale you had for having just foo() vs withFoo()? I see a few other large SDKs use setFoo() too.

Comment (conniey) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: I've updated the code. It does look a lot cleaner. :)

**Comment by conniey** [comment]
@JonathanGiles I've added an exception to be thrown if the ClientConfiguration is not valid (ie. they have not set credentials). How does it look now?
https://github.com/Azure/autorest-clientruntime-for-java/pull/592

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private class CredentialInformation {
+        private URL baseUri;
+        private String credential;
+        private byte[] secret;
+
+        URL baseUri() { return baseUri; }
+
+        String credential() { return credential; }
```
Comment (anuchandy) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: Is there a reason we call value of `id` segment in connection string as credentials? is it really credential or just application id?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/RequestIdPolicy.java [hunk 3]
```diff
@@ -12,7 +12,7 @@
 import java.util.UUID;
 
 /**
- * Creates a policy which puts a UUID in the request header. Azure uses
+ * Creates a addPolicy which puts a UUID in the request header. Azure uses
```
Comment (anuchandy) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/RequestIdPolicy.java]: nit; should we revert this change?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public ConfigurationClientCredentials(String connectionString) {
+        credentials = new CredentialInformation(connectionString);
+        headerProvider = new AuthorizationHeaderProvider(credentials);
+    }
+
+    URL baseUri() { return this.credentials.baseUri; }
+
+    @Override
+    public Mono<String> authorizationHeaderValueAsync(HttpRequest request) {
```
Comment (anuchandy) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: now I understand why we need `authorizationHeaderValueAsync(HttpRequest request)` instead of `authorizationHeaderValueAsync(String uri)`, sounds good

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationCredentialsPolicy.java [hunk 5]
```diff
@@ -90,28 +81,20 @@
                         return Mono.error(new NoSuchAlgorithmException("Unable to locate SHA-256 algorithm."));
                     }
 
-                    final Map<String, String> mapped = getDefaultHeaders(context.httpRequest().headers());
+                    final Map<String, String> mapped = getDefaultHeaders(context.httpRequest().url(), context.httpRequest().headers());
                     final String contentHash = Base64.getEncoder().encodeToString(messageDigest.digest());
 
                     mapped.put(CONTENT_HASH_HEADER, contentHash);
-
```
Comment (anuchandy) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationCredentialsPolicy.java]: In process method, we don't have to wait till `flatmap` to throw `NoSuchAlgorithmException`, we can use `Exceptions.propagate` in collect.supplier, like below. Also instead of creating new `NoSuchAlgorithmException` it will be good to rethrow it as it is.

```java
@Override
public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
    final Flux<ByteBuf> contents = context.httpRequest().body() == null
            ? Flux.just(getEmptyBuffer())
            : context.httpRequest().body().defaultIfEmpty(getEmptyBuffer());
    //
    return contents.collect(() -> {
        try {
            return MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw Exceptions.propagate(e);
        }
    }, (messageDigest, byteBuf) -> messageDigest.update(byteBuf.nioBuffer())).flatMap(messageDigest -> {
        final Map<String, String> mapped = getDefaultHeaders(context.httpRequest().url(), context.httpRequest().headers());
        final String contentHash = Base64.getEncoder().encodeToString(messageDigest.digest());

        mapped.put(CONTENT_HASH_HEADER, contentHash);
        mapped.forEach((key, value) -> context.httpRequest().headers().set(key, value));

        return next.process().doOnSuccess(this::logResponseDelegate);
    });
}
```

**Comment by conniey** [comment]
@anuchandy Thanks for all the good feedback. :D I'll start fixing them.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java (hunk 2 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: Hmm.. not sure. It existed before. I'll update it to id. From the docs, it looks like just the service identity for it.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+            String endpointString = "endpoint=";
+            String idString = "id=";
+            String secretString = "secret=";
+
+            for (String arg : args) {
+                String segment = arg.trim();
+                try {
+                    if (segment.toLowerCase().startsWith(endpointString)) {
+                        this.baseUri = new URL(segment.substring(segment.indexOf('=') + 1));
```
[the remark this answers is not in the document]
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: Nice catch! Fixed.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationCredentialsPolicy.java (hunk 5 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationCredentialsPolicy.java]: Thanks! I did not know that. Super helpful. Fixed.

**Comment by conniey** [comment]
@anuchandy @JonathanGiles Thanks for the suggestions/fixes. Any other glaring items? :)

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java [hunk 7]
```diff
@@ -55,38 +54,36 @@
...
-            List<HttpPipelinePolicy> policies = ConfigurationClient.getDefaultPolicies(connectionString);
-            policies.add(interceptorManager.getRecordPolicy());
-            policies.add(loggingPolicy);
-
-            pipeline = new HttpPipeline(HttpClient.createDefault().wiretap(true), policies);
+            client = ConfigurationClient.builder()
+                    .credentials(new ConfigurationClientCredentials(connectionString))
+                    .httpClient(HttpClient.createDefault().wiretap(true))
+                    .httpLogDetailLevel(HttpLogDetailLevel.BODY_AND_HEADERS)
+                    .addPolicy(interceptorManager.getRecordPolicy())
```
Comment (anuchandy) [inline applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: @JonathanGiles @conniey just double checking so we use `add**` prefix (e.g. addPolicy) as the general pattern to add items to collection configuration items?

**Comment by anuchandy** [comment]
@conniey LGTM, thanks!

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java (hunk 1 above)
Comment (anuchandy) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: same here, don't have a strong opinion on `foo(arg)` vs `withFoo(arg)`. I too liked `foo(arg)` so followed this in client-common.

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java (hunk 7 above)
Comment (JonathanGiles) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: I think that is best - thanks for double checking.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    .map(httpHeaders::value)
+                    .collect(Collectors.joining(";"));
+
+            // String-To-Sign=HTTP_METHOD + '\n' + path_and_query + '\n' + signed_headers_values
+            // Signed headers: "host;x-ms-date;x-ms-content-sha256"
+            return String.format("%s\n%s\n%s", request.httpMethod().toString().toUpperCase(Locale.US), pathAndQuery, signed);
+        }
+    }
+
+    private class CredentialInformation {
```
Comment (anuchandy) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: @conniey just a thought - when I looked closely on this type, this is actually an object representing `ConnectionString` right? so was wondering if so should it be named `ConnectionString` with a static factory method named `parse(string)` that takes connection string in string format and return `ConnectionString`? The usual pattern we see for connection string types.

Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: There are various formats for connectionstrings. We have an open issue to try and maybe make a ConnectionString some sort of ConnectionDetail. Or something. https://github.com/Azure/azure-sdk-for-java/issues/3080.

I can change this to be a factory method .parse for now.

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientCredentials.java]: For now just build what makes sense for Configuration Service, and don't invest too heavily in it - we will be changing this before too long, once the identity and authentication work lands as a common API.
